include/utilities/realtime/dataman/DataMan.h

```diff
@@ -0,0 +1,468 @@
+
+    inline bool isin(string a, json j)
+    {
+        for (unsigned int i = 0; i < j.size(); i++)
+        {
+            if (j[i] == a)
+                return true;
+        }
+        return false;
+    }
+
+    inline int closest(int v, json j, bool up)
+    {
+        int s = 100, k = 0, t;
+        for (unsigned int i = 0; i < j.size(); i++)
+        {
+            if (up)
+                t = j[i].get<int>() - v;
+            else
+                t = v - j[i].get<int>();
+            if (t >= 0 && t < s)
+            {
+                s = t;
+                k = i;
+            }
+        }
+        return k;
+    }
+
+    inline void check_shape(json &p_jmsg)
+    {
+        vector<size_t> varshape;
+        if (check_json(p_jmsg, {"varshape"}))
+        {
+            varshape = p_jmsg["varshape"].get<vector<size_t>>();
+        }
+        else
+        {
+            return;
+        }
+        if (p_jmsg["putshape"] == nullptr)
+        {
+            p_jmsg["putshape"] = varshape;
+        }
+        if (p_jmsg["offset"] == nullptr)
+        {
+            p_jmsg["offset"] = vector<size_t>(varshape.size(), 0);
+        }
+        p_jmsg["putbytes"] = product(p_jmsg["putshape"].get<vector<size_t>>(),
+                                     dsize(p_jmsg["dtype"].get<string>()));
+        p_jmsg["varbytes"] =
+            product(varshape, dsize(p_jmsg["dtype"].get<string>()));
+    }
+
+    inline shared_ptr<DataMan> get_man(string method)
+    {
+        string soname = "lib" + method + "man.so";
+        void *so = NULL;
+        so = dlopen(soname.c_str(), RTLD_NOW);
+        if (so)
+        {
+            shared_ptr<DataMan> (*func)() = NULL;
+            func = (shared_ptr<DataMan>(*)())dlsym(so, "getMan");
+            if (func)
+            {
+                return func();
+            }
+            else
+            {
+                logging("getMan() not found in " + soname);
+            }
+        }
+        else
+        {
+            logging("Dynamic library " + soname +
+                    " not found in LD_LIBRARY_PATH");
+        }
+        return nullptr;
+    }
+
+    string run_cmd(string cmd)
+    {
+        FILE *pipe = NULL;
+        char buffer[2048];
+        string result;
+        pipe = popen(cmd.c_str(), "r");
+        if (NULL == pipe)
+        {
+            perror("pipe");
+            return "";
+        }
+        while (!feof(pipe))
+        {
+            if (fgets(buffer, sizeof(buffer), pipe) != NULL)
+                result = buffer;
+        }
+        pclose(pipe);
+        return result;
+    }
+
+    std::function<void(const void *, string, string, string, vector<size_t>)>
+        m_callback;
+    map<string, shared_ptr<DataMan>> m_next;
+
+private:
+    json m_profiling;
+    chrono::time_point<chrono::system_clock> m_start_time;
+    chrono::time_point<chrono::system_clock> m_step_time;
+    bool m_profiling_enabled = false;
+};
```

Comment:
Since none of the functions involve templates then you can easily move thier implementation into a seperate cpp file.